### PR TITLE
Bicaws 1623 reject duplicate messages into bichard

### PR DESCRIPTION
**Pull request by emadkaramad**
In this PR:
- Added `messageHash` to `AuditLog` (constructor)
- Updated the code to pass `messageHash` to the `AuditLog` constructor
- Updated audit log API to validate message hash (whether it already exists)
- Updated `incoming-message-handler` to 
  - compute message hash(SHA256) from the inner message XML(Content inside `<DataStreamContent>` element)
  - add a new failure step to the state machine if the message hash already exists
- Added `getMessageByHash` to `AuditLogApiClient`
- Added `fetchByHash` to `AuditLogDynamoGateway`
- Updated Audit Log API Client to return our error messages instead of axios errors 

also,
- Removed the retry logic from the event handler as we now use the built-in step function error handler.
- Enabled validation for `s3Path`, `exeternalId`, and `stepExecutionId` fields in `audit-log-api` when creating a new audit log item.

### CODE REVIEW

**Comment by bjpirt** [comment]
I was wondering if it was worth doing the validation slightly differently?

At the moment this is making an HTTP request to the API the check the hash and then making an HTTP request to create the message. We could make the Create Message endpoint check the unique hash doesn't exist and return an error if it does, that way we'd only have one HTTP request

File: src/audit-log-api/src/handlers/createAuditLog.test.ts [hunk 1]
```diff
@@ -1,14 +1,18 @@
+jest.mock("src/use-cases/validateCreateAuditLog")
```
Comment (tomvaughan77) [inline src/audit-log-api/src/handlers/createAuditLog.test.ts:1]: Nit: This feels like an odd place to put a mock? I'd normally expect imports to be the first thing in a file?

File: src/incoming-message-handler/src/handlers/storeMessage.ts [hunk 2]
```diff
@@ -1,29 +1,42 @@
...
   const receivedMessage = await retrieveMessageFromS3(event, s3Gateway)
 
   if (isError(receivedMessage)) {
     throw receivedMessage
   }
 
-  if ("validationResult" in receivedMessage) {
-    return receivedMessage
+  if ("isValid" in receivedMessage) {
+    logger.info(JSON.stringify(receivedMessage))
```
Comment (tomvaughan77) [inline src/incoming-message-handler/src/handlers/storeMessage.ts:38]: Nit: Should this be `info` or `debug`?

File: src/audit-log-api/src/handlers/createAuditLog.test.ts (hunk 1 above)
Comment (emadkaramad) [reply src/audit-log-api/src/handlers/createAuditLog.test.ts]: We had an issue with mocking modules when we used `jest.mock()` after imports previously. However, I put it after imports in this test and it seems to be okay. 🤷‍♂️

Updated.

File: src/incoming-message-handler/src/handlers/storeMessage.ts (hunk 2 above)
Comment (emadkaramad) [reply src/incoming-message-handler/src/handlers/storeMessage.ts:38]: This log here is not for debugging purposes. We are logging to cloud watch here to be able to use this info in monitoring and alerting later.

File: src/shared/src/AuditLogDynamoGateway/AwsAuditLogDynamoGateway.integration.test.ts [hunk 3]
```diff
@@ -500,15 +501,51 @@ describe("AuditLogDynamoGateway", () => {
...
+        [...Array(3).keys()].map(async (i: number) => {
+          const auditLog = new AuditLog(`External correlation id ${i}`, new Date(), `hash-${i}`)
+          await gateway.create(auditLog)
+        })
+      )
+
+      const hash = "hash-2"
+      const result = await gateway.fetchByHash(hash)
+
+      console.log(result)
```
Comment (tomvaughan77) [inline src/shared/src/AuditLogDynamoGateway/AwsAuditLogDynamoGateway.integration.test.ts:516]: Nit: Leftover logging?

Comment (emadkaramad) [reply src/shared/src/AuditLogDynamoGateway/AwsAuditLogDynamoGateway.integration.test.ts]: Good spot! Removed.

**Review by tomvaughan77 — DISMISSED** [review]

**Review by bjpirt — APPROVED** [review]

**Review by bjpirt — APPROVED** [review]

**Pull request merged**